core/services/ocr3/plugins/ccip/commit/plugin.go

```diff
@@ -3,115 +3,594 @@ package commit
 	for _, ao := range aos {
-		parsedObservation, err := model.DecodeCommitPluginObservation(ao.Observation)
+		obs, err := model.DecodeCommitPluginObservation(ao.Observation)
 		if err != nil {
-			p.lggr.Errorw("decode commit plugin observation", "err", err)
-			return ocr3types.Outcome{}, err
+			return ocr3types.Outcome{}, fmt.Errorf("decode commit plugin observation: %w", err)
 		}
-		msgsFromObservations = append(msgsFromObservations, parsedObservation.NewMsgs...)
+		decodedObservations = append(decodedObservations, obs)
 	}
 
-	sourceChains, groupedMsgs := slicelib.GroupBy(
-		msgsFromObservations, func(msg model.CCIPMsgBaseDetails) model.ChainSelector { return msg.SourceChain })
-	for _, sourceChain := range sourceChains {
-		p.lggr.Debugf("for source chain %d we got %d msg observations", len(groupedMsgs[sourceChain]))
+	maxSeqNumsConsensus, err := p.maxSeqNumsConsensus(decodedObservations)
+	if err != nil {
+		return ocr3types.Outcome{}, fmt.Errorf("max sequence numbers consensus: %w", err)
+	}
+	p.lggr.Debugw("max sequence numbers consensus", "maxSeqNumsConsensus", maxSeqNumsConsensus)
+
+	merkleRoots, err := p.newMsgsConsensus(maxSeqNumsConsensus, decodedObservations)
+	if err != nil {
+		return ocr3types.Outcome{}, fmt.Errorf("new messages consensus: %w", err)
 	}
+	p.lggr.Debugw("new messages consensus", "merkleRoots", merkleRoots)
 
-	return ocr3types.Outcome{}, fmt.Errorf("implement me")
+	return model.NewCommitPluginOutcome(maxSeqNumsConsensus, merkleRoots).Encode()
 }
 
 func (p *Plugin) Reports(seqNr uint64, outcome ocr3types.Outcome) ([]ocr3types.ReportWithInfo[[]byte], error) {
-	/*
-		# Only a single report is needed containing a batch of gas price updates for the chain.
-		# and a list of roots by source. We only include the gas price batch update
-		# if a timer has expired - ie periodically we batch write all gas prices.
-	*/
-	panic("implement me")
+	outc, err := model.DecodeCommitPluginOutcome(outcome)
+	if err != nil {
+		return nil, fmt.Errorf("decode commit plugin outcome: %w", err)
+	}
+
+	// todo: include token price updates
+	// todo: include gas price updates
+	priceUpdates := make([]model.TokenPriceUpdate, 0)
+
+	rep := model.NewCommitPluginReport(outc.MerkleRoots, priceUpdates)
+	encodedReport, err := p.reportCodec.Encode(context.Background(), rep)
+	if err != nil {
+		return nil, fmt.Errorf("encode commit plugin report: %w", err)
+	}
+
+	return []ocr3types.ReportWithInfo[[]byte]{{Report: encodedReport, Info: nil}}, nil
 }
 
 func (p *Plugin) ShouldAcceptAttestedReport(ctx context.Context, u uint64, r ocr3types.ReportWithInfo[[]byte]) (bool, error) {
-	panic("implement me")
+	decodedReport, err := p.reportCodec.Decode(ctx, r.Report)
+	if err != nil {
+		return false, fmt.Errorf("decode commit plugin report: %w", err)
+	}
+
+	isEmpty := decodedReport.IsEmpty()
+	if isEmpty {
+		p.lggr.Infow("skipping empty report")
+		return false, nil
+	}
+
+	return true, nil
```

Comment:
Could be simplified to:
```suggestion
	return !decodedReport.IsEmpty(), nil
```

But I guess we'd lose the log, so no biggie.

Reply:
yeah i wanted that log